include/pybind11/stl.h

```diff
@@ -35,6 +36,74 @@
 PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
 PYBIND11_NAMESPACE_BEGIN(detail)
 
+//
+// Begin: Equivalent of
+//        https://github.com/google/clif/blob/337b2a106c0552ad85d40cfc3797465756840ea0/clif/python/runtime.cc#L388-L424
+/*
+The three `PyObjectTypeIsConvertibleTo*()` functions below are
+the result of converging the behaviors of pybind11 and PyCLIF
+(http://github.com/google/clif).
+
+Originally PyCLIF was extremely far on the permissive side of the spectrum,
+while pybind11 was very far on the strict side. Originally PyCLIF accepted any
+Python iterable as input for a C++ `vector`/`set`/`map` argument, as long as
+the elements were convertible. The obvious (in hindsight) problem was that
+any empty Python iterable could be passed to any of these C++ types, e.g. `{}`
+was accpeted for C++ `vector`/`set` arguments, or `[]` for C++ `map` arguments.
+
+The functions below strike a practical permissive-vs-strict compromise,
+informed by tens of thousands of use cases in the wild. A main objective is
+to prevent accidents and improve readability:
+
+- Python literals must match the C++ types.
+
+- For C++ `set`: The potentially reducing conversion from a Python sequence
+  (e.g. Python `list` or `tuple`) to a C++ `set` must be explicit, by going
+  through a Python `set`.
+
+- However, a Python `set` can still be passed to a C++ `vector`. The rationale
+  is that this conversion is not reducing. Implicit conversions of this kind
+  are also fairly commonly used, therefore enforcing explicit conversions
+  would have an unfavorable cost : benefit ratio; more sloppily speaking,
+  such an enforcement would be more annyoing than helpful.
+*/
+
+inline bool PyObjectIsInstanceWithOneOfTpNames(PyObject *obj,
+                                               std::initializer_list<const char *> tp_names) {
+    if (PyType_Check(obj)) {
+        return false;
+    }
+    const char *obj_tp_name = Py_TYPE(obj)->tp_name;
+    for (const auto *tp_name : tp_names) {
+        if (strcmp(obj_tp_name, tp_name) == 0) {
+            return true;
+        }
+    }
+    return false;
+}
+
+inline bool PyObjectTypeIsConvertibleToStdVector(PyObject *obj) {
+    if (PySequence_Check(obj) != 0) {
+        return !PyUnicode_Check(obj) && !PyBytes_Check(obj);
+    }
+    return (PyGen_Check(obj) != 0) || (PyAnySet_Check(obj) != 0)
+           || PyObjectIsInstanceWithOneOfTpNames(
+               obj, {"dict_keys", "dict_values", "dict_items", "map", "zip"});
+}
+
+inline bool PyObjectTypeIsConvertibleToStdSet(PyObject *obj) {
+    return (PyAnySet_Check(obj) != 0) || PyObjectIsInstanceWithOneOfTpNames(obj, {"dict_keys"});
+}
+
+inline bool PyObjectTypeIsConvertibleToStdMap(PyObject *obj) {
+    return (PyDict_Check(obj) != 0)
+           || ((PyMapping_Check(obj) != 0) && (PyObject_HasAttrString(obj, "items") != 0));
```

Comment:
We need to check the signature of items or that attrs is at least callable, right? Otherwise, it will try (and fail to convert a class with an items attr attached to it, right?

Reply:
Good point, I didn't consider this before. PyCLIF doesn't have function overloading, it does not matter there.

Adding the callable check is easy (done), but I don't know if there is a way to check if the callable can be called without arguments (without actually calling it).

The conditions under which the missing check for number of arguments could go wrong are very special:

* obj needs to have `__getitem__` (to pass `PyMapping_Check()`)
* obj needs to have `items` and `items` must be callable
* there must be at least two overloads for the function for which we're converting from Python to C++
* one of the overloads must expect `std::map` (or similar)
* another the wrapped C++ type in the same position
* the overloads must be ordered so that the `std::map` conversion is tried first

Only then will the desired overload will not be reached.

Even if we can check for the number of arguments, `.items()` could have some arbitrary return value and we'd fall over processing that.

The only truly certain alternative is to not have `PyMapping_Check()` at all.

From memory: not having the `PyMapping_Check()` + `.items()` path would not be easy to accommodate. I think I'd have to make pretty intrusive changes to user code.

On the other hand, having the requirement that an object with `.__getitem__()` & `.items()` must implement these to be compatible with a Python mapping seems very natural (what most everybody would expect and not easily violate). What do you think?

But in any event, adding the callable check is easy and I'm happy to keep that.

Reply:
I just discovered `PyMapping_Items()`: 4ab40d74cb4d612d37ef196c70721df92a0be665

Using that makes the "mapping" requirement more formal.

Our gating conditions are stronger than those for `PyMapping_Check()`, I think supporting Python mappings in this way is a great practical compromise, usefulness vs purity.